### PR TITLE
Feature/145081 msgpack

### PR DESCRIPTION
**Pull request by vananyev**

### CODE REVIEW

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

File: test/msgpack4s/primitives4s.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
+package msgpack4s
```
Comment (vananyev) [inline test/msgpack4s/primitives4s.scala:1]: This is a bunch of unit tests what has to be reviewed

File: tools/msgpack4s/src/main/scala/org.velvia/msgpack/ExtraCodecs.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  // BigDecimals are packed as an 'Int' for 'scale' value and as 'ByteArray' for BigDecimal value
+  // I.e. a codec appears packed 'Int', it assums it like BigDecimal value and unpacks next unscaledValue as 'ByteArray'
+  implicit object BigDecimalCodec extends Codec[BigDecimal] {
+
+    // Packing section
+    def pack(out: DataOutputStream, bd: BigDecimal) {
+      IntCodec.pack(out, bd.scale)
+      ByteArrayCodec.pack(out, bd.unscaledValue().toByteArray)
+    }
+
```
Comment (vananyev) [inline tools/msgpack4s/src/main/scala/org.velvia/msgpack/ExtraCodecs.scala:19]: Bellow part (next 11 lines) is a fix of BigDecimal deserialization, what has to be reviewed.

File: test/msgpack4s/primitives4s.scala [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      import org.json4s.JsonAST._
+      import org.velvia.msgpack.Json4sCodecs._
+
+      val aray = parse("""[1, 2.5, null, "Streater"]""")
+      val map = parse("""{"bool": true, "aray": [3, -4], "map": {"inner": "me"}}""")
+
+      unpack[JArray](pack(aray)) mustBe aray
+      unpack[JValue](pack(aray)) mustBe aray
+      unpack[JValue](pack(map)) mustBe map
+    }
```
Comment (dmitriyisaev) [inline test/msgpack4s/primitives4s.scala:43]: Should we assert every array/map member or at least size?

File: test/msgpack4s/primitives4s.scala [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      import org.velvia.msgpack
+      import play.api.libs.json.{JsValue, Json}
+      import org.velvia.msgpack.PlayJsonCodecs.JsValueCodec
+
+      val js: JsValue = Json.parse("""{"amount":40.1,"currency":"USD","label":"10.00"}""")
+
+      val p: Array[Byte] = msgpack.pack(js)
+
+      val u = msgpack.unpack[JsValue](p)
+    }
```
Comment (dmitriyisaev) [inline test/msgpack4s/primitives4s.scala]: Did we missed assert?

File: tools/msgpack4s/src/main/scala/org.velvia/msgpack/ExtraCodecs.scala (hunk 2 above)
Comment (dmitriyisaev) [reply tools/msgpack4s/src/main/scala/org.velvia/msgpack/ExtraCodecs.scala:19]: Seems fine for me. Not scala way but ok for codec (It's implemented in the same style as source library).

File: test/msgpack4s/primitives4s.scala [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      import org.velvia.msgpack
+      import org.velvia.msgpack.Json4sCodecs.JValueCodec
+      import org.json4s.native.JsonMethods.parse
+      import org.json4s.JsonAST.JValue
+
+
+      val js: JValue = parse("""{"amount":40.1,"currency":"USD","label":"10.00"}""")
+      var p: Array[Byte] = msgpack.pack(js)
+      val u = msgpack.unpack[JValue](p)
+    }
```
Comment (dmitriyisaev) [inline test/msgpack4s/primitives4s.scala]: Did we missed assert?

File: test/msgpack4s/primitives4s.scala (hunk 3 above)
Comment (vananyev) [reply test/msgpack4s/primitives4s.scala:43]: No, existed asserts are anought, bcs there are comaprations btw packend and unpacked data with original values, i.e.: "unpack[JArray](pack(aray)) mustBe aray" - aray is the data.

File: test/msgpack4s/primitives4s.scala (hunk 4 above)
Comment (vananyev) [reply test/msgpack4s/primitives4s.scala]: Yes, assert were mixed. Fixed.

File: test/msgpack4s/primitives4s.scala (hunk 5 above)
Comment (vananyev) [reply test/msgpack4s/primitives4s.scala]: Yes, assert were mixed. Fixed.

**Comment by vananyev** [comment]
@dmitriyisaev , i have added the comments and asserts where you asked. Please check.

**Review by dmitriyisaev — APPROVED** [review]
Looks fine for me

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Pull request merged**